docs/proposals/PROPOSAL_D.md

```diff
@@ -0,0 +1,413 @@
+        "mpi.version": "4.1.1",
+        "os.name": "Ubuntu 22.04.3 LTS",
+        "os.release": "22.04.3",
+        "os.vendor": "ubuntu",
+        "os.version": "22.04",
+        "hardware.gpu.available": "yes"
+      }
+    }
+  ]
+}
+```
+
+Note that I've also added the schema to be alongside the different sections.
+
+## Frequently Asked Questions
+
+### Is a compatibility interest group required to define relationships?
+
+No. A schema without relationshiops is essentially defining a namespace and dictionary of metadata attributes. Some underlying tool would not find edges and would take a more "bag of features" approach to assess compatibility.
+
+### How would graphs be combined?
+
+This brings up the question of how to handle two separate graphs. I don't think there can be dependencies between the two (e.g., a reference to archspec from supercontainers) because that introduces the potential for too much error. Rather, I think we should allow the selection process to move down the list of annotations, take one of the following approaches:
+
+1. Ranked: Go down the list, and honor the order. This means honoring graph selection and bag of metadata feature selection for the first, and then providing a subset of images to the second.
+2. Graphs first, then features: This means honoring graph selection of the first, then graph selection of the second, and putting all the final contenders into one bag of (combined) features.
+
+Likely, the second provides a more holistic view. In practice for the second point, and with archpsec and supercontainers, this would mean (for a parsing process):
+
+1. start with an entire set of images and matched compatibility artifacts
+2. filter down to a subgraph based on archspec (arm64)
+3. give this new subset of images to a second graph (supercontainers)
+4. allow supercontainers to filter down to the set they care about.
+
+I am going to attempt prototyping this soon.
+
+## Compatibility Plugins
+
+Note that the previous plugin design is still relevant, and now I have a [prototype](https://github.com/supercontainers/compspec-go/blob/main/docs/usage) that demonstrates some of the ideas. I will be updating the prototype with this new design.
+
+## Example Algorithm
+
+As an example, let's consider first a set of flat annotations that include a higher level architecture and then a library version. These would be subsets of compatibility artifacts defined by Proposal C, meaning that they can be mapped into a graph format, but the artifact format does not require it and is presented simply as the list. For this example, the extraction (creation of the artifacts) can be assumed to be done by a community that cares about architecture, cpu, and an MPI variant. The OCI working group should not control or bias what can be done with the artifacts generated in this regard. Below we have five slices of artifacts (json objects) that describe a set of corresponding 5 container images.
```

Comment:
> meaning that they can be mapped into a graph format, but the artifact format does not require it and is presented simply as the list.

Unfortunately there are cases where it will require it. 
Now I understand the major difference in mine and your flows. 
I talk about graph in context of compatibility data. You're talking about graph for schema itself.

You might have cases where in the bag of features you have to implement OR and that can grow into graph.
A very simple example where you need OR in the compatbility itself:

```json
{
  "compatibilities": [
    {
      "name": "org.supercontainers",
      "version": "0.0.0",
      "annotations": {
        "hardware.cpu.optimized.mode": "generic",
        "hardware.cpu.optimized.version": "skylake"
      }
    },
    {
      "name": "archspec.io",
      "version": "0.0.0",
      "annotations": {
        "name": "amd64"
      }
    },
    {
      "name": "io.gpu",
      "version": "0.0.0",
      "annotations": {
        "vendor": "NVIDIA"
       }
     },
     {
      "name": "io.gpu",
      "version": "0.0.0",
      "annotations": {
        "vendor": "Intel"
       }
     }
  ]
}
```

One container image can be compatible with different PCI vendors. In the above example only one compatibilitiy of `io.gpu` will pass, the other will fail, but still the result should be reported that the container is compatible with the host if at least one passes.

The complexity grows if you include other stuff, like kernel modules or cmdline parameters based on CPU etc. and all of that can be around one container image, meaning one compatibility artifact.

Reply:
That's a good point - in the above example we would need OR represented as an edge between two things (likely the same type of GPU). I think my proposal would support that, but there would need to be implicit OR (and other) relationships added in some respect to the artifact that the user requests. And yeah, that's the comment I alluded to above - we are in agreement!